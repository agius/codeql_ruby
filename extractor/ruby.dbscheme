/** dbscheme for Ruby, feel free to edit. I don't know what I'm doing ^_^; */

sourceLocationPrefix(varchar(900) prefix : string ref);

leaf_nodes(unique int id: @leaf_node, string text: string ref, int beginLine: int ref, int beginColumn: int ref)
