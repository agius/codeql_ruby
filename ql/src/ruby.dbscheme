/** dbscheme for Ruby, feel free to edit. I don't know what I'm doing ^_^; */

/* Filesystem fragments specified as part of the extractor contract */

sourceLocationPrefix(varchar(900) prefix : string ref);

@location = @location_default /* TODO add alternatives */;

locations_default(
  unique int id:    @location_default,
  int file:         @file ref,
  int beginLine:    int ref,
  int beginColumn:  int ref,
  int endLine:      int ref,
  int endColumn:    int ref
);

has_location(
  unique int locatable: @locatable ref,
  int location:         @location ref
);

@sourceline = @file /* TODO add alternatives */;

numlines(
  int element_id:   @sourceline ref,
  int num_lines:    int ref,
  int num_code:     int ref,
  int num_comment:  int ref
);

/*
  fromSource(0) = unknown,
  fromSource(1) = from source,
  fromSource(2) = from library
*/
files(
  unique int id:        @file,
  varchar(900) name:    string ref,
  varchar(900) simple:  string ref,
  varchar(900) ext:     string ref,
  int fromSource:       int ref
);

folders(
  unique int id:        @folder,
  varchar(900) name:    string ref,
  varchar(900) simple:  string ref
);

@container = @folder | @file ;

containerparent(
  int parent:       @container ref,
  unique int child: @container ref
);

/* Ruby AST Stuff */

leaf_nodes(
  unique int id:    @leaf_node,
  string text:      string ref,
  int beginLine:    int ref,
  int beginColumn:  int ref
);

@locatable = @leaf_node;
