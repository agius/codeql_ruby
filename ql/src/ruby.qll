
import Files
import Locations
import LeafNode
