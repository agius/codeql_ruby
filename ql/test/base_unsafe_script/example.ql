/**
 * @name Script contains anything
 * @kind problem
 * @problem.severity warning
 * @id ruby/spec/base-unsafe-script
 */

import ruby

from LeafNode n
select n.getText(), "This is a leaf node."
