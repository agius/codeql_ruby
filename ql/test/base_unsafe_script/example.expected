| 1 | This is a leaf node. |
| ARGV | This is a leaf node. |
| eval | This is a leaf node. |
